prow/cmd/tide/main.go

```diff
@@ -174,33 +176,22 @@ func main() {
 	server := &http.Server{Addr: ":" + strconv.Itoa(o.port)}
 
 	// Push metrics to the configured prometheus pushgateway endpoint or serve them
-	metrics.ExposeMetrics("tide", cfg().PushGateway)
+	metrics.ExposeMetrics("tide", cfg().PushGateway, mgr)
 
 	start := time.Now()
 	sync(c)
 	if o.runOnce {
-		return
+		os.Exit(0)
 	}
-	go func() {
-		sig := make(chan os.Signal, 1)
-		signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
-		for {
-			select {
-			case <-time.After(time.Until(start.Add(cfg().Tide.SyncPeriod.Duration))):
-				start = time.Now()
-				sync(c)
-			case <-sig:
-				logrus.Info("Tide is shutting down...")
-				// Shutdown the http server with a 10s timeout then return to execute
-				// deferred c.Shutdown()
-				ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
-				defer cancel() // frees ctx resources
-				server.Shutdown(ctx)
-				return
-			}
-		}
-	}()
-	logrus.WithError(server.ListenAndServe()).Warn("Tide HTTP server stopped.")
+
+	// serve data
+	mgr.ListenAndServe(server, 10*time.Second)
+
+	// run the controller, but only after one sync period expires after our first run
+	time.Sleep(time.Until(start.Add(cfg().Tide.SyncPeriod.Duration)))
+	mgr.Tick(func() {
+		sync(c)
+	}, cfg().Tide.SyncPeriod.Duration)
```

Comment:
This change makes Tide require a restart to detect changes to the sync period config option.
It would probably be preferable for the `Tick` function to take a `func() time.Duration` so that the tick duration can change between ticks.

Reply:
How would we structure the internals of `Tick()` then?

Reply:
Like, we would need one thread watching the output of the `func() time.Duration` to see if it's changed, then, if it has, change the `time.Tick()` impl feeding the channel used in the ticker? hrm